File: 0143-reorder-list/0143-reorder-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {void} Do not return anything, modify head in-place instead.
 */
var reorderList = function(head) {
    
    if (!head)
        return;
    
    if (!head.next)
        return head;
    
    const stack = []
    let node = head
    length = 0

    while(node) {
        stack.push(node)
        node = node.next
        length += 1
    }

    node = head
        
    console.log(length, stack, Math.floor(stack.length/2))

    
    
    for (let i = 0; i < Math.floor(length/2); i += 1)
    {
        console.log(i)
        getNode = stack.pop();
        console.log(11)
        nextNode = node.next;
        console.log(22)

        node.next = getNode;
        console.log(33)
        getNode.next = nextNode;
        console.log(44)
        node = nextNode;
        console.log(55)
    }
    
    // if node.next
    if (length % 2 > 0) {
        node.next = null
    } else {
        node.next.next = null
    }

};

File: 0143-reorder-list/0143-reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if not head:
            return
        
        stack = []
        node = head
        
        while node:
            stack.append(node)
            node = node.next
        
        node = head
        
        for i in range(len(stack)//2):
            getNode = stack.pop()
            nextNode = node.next
            node.next = getNode
            getNode.next = nextNode
            node = nextNode
        
        node.next = None    
